# 03 — Deep Dive: Growth, Change and Earnings Quality

Notebook 02 shows the levels. This one asks three questions the levels do not
answer:

1. **How fast is each business actually growing?** Margin tells you nothing about
   scale.
2. **Which direction is each metric moving, and by how much?** Year-on-year
   changes in percentage *points*, not percentages of percentages.
3. **Is the reported profit real?** Cash conversion — operating cash flow over net
   income — is the check on earnings quality.

In [1]:
import sys

sys.path.append("..")

import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots

import config

fundamentals = pd.read_pickle("../data/fundamentals.pkl")
YEARS = config.ANALYSIS_YEARS
COLORS = config.COLORS

print(fundamentals[["company", "year", "revenue", "net_income",
                    "operating_cash_flow", "free_cash_flow"]]
      .set_index(["company", "year"]).div(1e9).round(2).to_string())

              revenue  net_income  operating_cash_flow  free_cash_flow
company year                                                          
Tesla   2022    81.46       12.56                14.72            7.57
        2023    96.77       15.00                13.26            4.36
        2024    97.69        7.09                14.92            3.58
BYD     2022    63.02        2.47                20.93            6.45
        2023    85.17        4.25                24.00            6.74
        2024   108.14        5.60                18.57            5.02
Ford    2022   158.06       -1.98                 6.85           -0.01
        2023   176.19        4.35                14.92            6.68
        2024   184.99        5.88                15.42            6.74


## Section 1 — Revenue growth

Absolute scale and growth rate, which the ratio metrics hide entirely.

In [2]:
revenue = fundamentals.pivot_table(index="year", columns="company",
                                   values="revenue", observed=True) / 1e9
growth = revenue.pct_change() * 100

print("Revenue ($B):")
print(revenue.round(1).to_string())
print("\nYoY growth (%):")
print(growth.round(1).to_string())

first, last = config.FIRST_YEAR, config.LAST_YEAR
print(f"\nTotal growth {first}-{last}:")
for name in config.COMPANIES:
    change = (revenue.loc[last, name] / revenue.loc[first, name] - 1) * 100
    print(f"  {name:6s} ${revenue.loc[first, name]:6.1f}B -> ${revenue.loc[last, name]:6.1f}B  ({change:+.1f}%)")

Revenue ($B):
company  Tesla    BYD   Ford
year                        
2022      81.5   63.0  158.1
2023      96.8   85.2  176.2
2024      97.7  108.1  185.0

YoY growth (%):
company  Tesla   BYD  Ford
year                      
2022       NaN   NaN   NaN
2023      18.8  35.2  11.5
2024       0.9  27.0   5.0

Total growth 2022-2024:
  Tesla  $  81.5B -> $  97.7B  (+19.9%)
  BYD    $  63.0B -> $ 108.1B  (+71.6%)
  Ford   $ 158.1B -> $ 185.0B  (+17.0%)


In [3]:
fig_growth = go.Figure()
for name in config.COMPANIES:
    fig_growth.add_trace(go.Bar(
        x=growth.index.astype(str),
        y=growth[name],
        name=name,
        marker_color=COLORS[name],
    ))
fig_growth.update_layout(
    title_text="Revenue growth, year on year (%)",
    template="plotly_white",
    barmode="group",
    height=400,
    width=800,
    yaxis_title="YoY growth (%)",
)
fig_growth.show()

**BYD grew revenue roughly four times as fast as either of the others across the
window** — +72% against +20% and +17% — and in 2024 alone it grew 27% while Tesla
grew 0.9%. All three decelerated into 2024; the gap is in how far from zero each
landed. Tesla's near-flat 0.9% is the fact its 189x multiple has to be argued
against.

## Section 2 — Year-on-year change

Changes in ratio metrics are expressed in **percentage points**, because a move
from 15.4% to 7.3% is a fall of 8.2 points, not 8.2 per cent. Conflating the two
is a common way to state a number that is wrong by a factor of two.

In [4]:
CHANGE_METRICS = [("npm", "Net Profit Margin (pp)"),
                  ("roa", "Return on Assets (pp)"),
                  ("free_cash_flow", "Free Cash Flow ($B)")]

changes = {}
for field, label in CHANGE_METRICS:
    wide = fundamentals.pivot_table(index="year", columns="company",
                                    values=field, observed=True)
    if field == "free_cash_flow":
        wide = wide / 1e9
    changes[label] = wide.diff().dropna()

for label, frame in changes.items():
    print(f"\n{label} — change vs prior year:")
    print(frame.round(2).to_string())


Net Profit Margin (pp) — change vs prior year:
company  Tesla   BYD  Ford
year                      
2023      0.08  1.07  3.72
2024     -8.24  0.19  0.71

Return on Assets (pp) — change vs prior year:
company  Tesla   BYD  Ford
year                      
2023     -1.18  0.96  2.36
2024     -8.26  0.78  0.47

Free Cash Flow ($B) — change vs prior year:
company  Tesla   BYD  Ford
year                      
2023     -3.21  0.29  6.70
2024     -0.78 -1.71  0.06


In [5]:
fig_yoy = make_subplots(rows=1, cols=3, subplot_titles=[label for _, label in CHANGE_METRICS])

for col, (_, label) in enumerate(CHANGE_METRICS, start=1):
    frame = changes[label]
    for name in config.COMPANIES:
        fig_yoy.add_trace(go.Bar(
            x=frame.index.astype(str),
            y=frame[name],
            name=name,
            marker_color=COLORS[name],
            legendgroup=name,
            showlegend=(col == 1),
        ), row=1, col=col)

fig_yoy.update_layout(
    title_text="Year-on-year change (percentage points, and $B for cash flow)",
    template="plotly_white",
    barmode="group",
    height=420,
    width=1200,
)
fig_yoy.show()

The two-year moves that the theses in notebook 04 rest on:

| | Tesla | BYD | Ford |
|---|---:|---:|---:|
| Net profit margin | **−8.2 pp** (15.4 → 7.3) | +1.3 pp (3.9 → 5.2) | +4.4 pp (−1.3 → 3.2) |
| Return on assets | **−9.4 pp** (15.2 → 5.8) | +1.7 pp (3.5 → 5.2) | +2.8 pp (−0.8 → 2.1) |

Tesla's return on assets falls further than its margin — 9.4 points against 8.2 —
because the asset base grew 48% over the same window while net income fell 44%.
The margin compression is the story people tell; the asset build underneath it is
what makes the return fall faster.

## Section 3 — Earnings quality

**Cash Conversion Ratio = Operating Cash Flow / Net Income.**

Below 1.0 a company is booking more profit than it collects in cash, which is the
classic earnings-quality warning. Well above 1.0 is normal for
capital-intensive manufacturers, where large non-cash depreciation charges sit
between operating cash and reported profit.

The ratio is undefined in any useful sense when net income is negative — Ford's
2022 reading of −3.46 divides positive cash flow by a loss and should be read as
"not meaningful", not as a number on the same scale as the others.

In [6]:
ccr = fundamentals.pivot_table(index="year", columns="company", values="ccr", observed=True)
print("Cash Conversion Ratio (OCF / Net Income):")
print(ccr.round(2).to_string())
print("\nNote: Ford 2022 divides by a net loss — not meaningful.")

fig_ccr = go.Figure()
for name in config.COMPANIES:
    series = ccr[name]
    if name == "Ford":
        series = series[series.index > config.FIRST_YEAR]  # drop the loss-year artefact
    fig_ccr.add_trace(go.Scatter(
        x=series.index, y=series.values, name=name, mode="lines+markers",
        line=dict(color=COLORS[name], width=2), marker=dict(size=8),
    ))
fig_ccr.add_hline(y=1.0, line_dash="dash", line_color="grey",
                  annotation_text="1.0 — profit fully converted to cash")
fig_ccr.update_layout(
    title_text="Cash Conversion Ratio (Ford 2022 excluded — divides by a loss)",
    template="plotly_white", height=420, width=850,
    yaxis_title="OCF / Net Income", hovermode="x unified",
)
fig_ccr.update_xaxes(tickmode="array", tickvals=YEARS, ticktext=[str(y) for y in YEARS])
fig_ccr.show()

Cash Conversion Ratio (OCF / Net Income):
company  Tesla   BYD  Ford
year                      
2022      1.17  8.47 -3.46
2023      0.88  5.65  3.43
2024      2.10  3.32  2.62

Note: Ford 2022 divides by a net loss — not meaningful.


### Reading the cash conversion

- **Tesla** dipped below 1.0 in 2023 (0.88) — the one year in this window where a
  company in the set reported more profit than it generated in cash — then
  recovered to 2.10 in 2024. The recovery is not a strengthening business: net
  income fell 53% while operating cash flow rose 13%, so most of the move came
  from the denominator collapsing rather than the numerator improving. Tesla's 2.10 is the **lowest** of the three
  in 2024, though still comfortably above 1.
- **BYD** is the highest at 3.32, but it has fallen every year from 8.47. That
  decline is not deterioration: it is net income growing far faster (+127% over
  the window) than operating cash flow (−11%). A CCR falling towards a normal
  manufacturing range from an extreme starting point is convergence, not a warning.
- **Ford** sits at 2.62, between the other two, on a genuine recovery from a 2022
  loss.

No company in this set fails the earnings-quality test in 2024. The one reading
that ever did — Tesla's 0.88 in 2023 — has reversed.